lib/src/merge.rs

```diff
@@ -92,48 +98,48 @@ mod tests {
         assert_eq!(trivial_merge(&[0, 0], &[0, 0, 0]), Some(&0));
         assert_eq!(trivial_merge(&[0, 0], &[0, 0, 1]), Some(&1));
         assert_eq!(trivial_merge(&[0, 0], &[0, 1, 0]), Some(&1));
-        assert_eq!(trivial_merge(&[0, 0], &[0, 1, 1]), None);
+        assert_eq!(trivial_merge(&[0, 0], &[0, 1, 1]), Some(&1));
         assert_eq!(trivial_merge(&[0, 0], &[0, 1, 2]), None);
         assert_eq!(trivial_merge(&[0, 0], &[1, 0, 0]), Some(&1));
-        assert_eq!(trivial_merge(&[0, 0], &[1, 0, 1]), None);
+        assert_eq!(trivial_merge(&[0, 0], &[1, 0, 1]), Some(&1));
         assert_eq!(trivial_merge(&[0, 0], &[1, 0, 2]), None);
-        assert_eq!(trivial_merge(&[0, 0], &[1, 1, 0]), None);
+        assert_eq!(trivial_merge(&[0, 0], &[1, 1, 0]), Some(&1));
         assert_eq!(trivial_merge(&[0, 0], &[1, 1, 1]), Some(&1));
         assert_eq!(trivial_merge(&[0, 0], &[1, 1, 2]), None);
         assert_eq!(trivial_merge(&[0, 0], &[1, 2, 0]), None);
         assert_eq!(trivial_merge(&[0, 0], &[1, 2, 1]), None);
         assert_eq!(trivial_merge(&[0, 0], &[1, 2, 2]), None);
         assert_eq!(trivial_merge(&[0, 0], &[1, 2, 3]), None);
-        assert_eq!(trivial_merge(&[0, 1], &[0, 0, 0]), None);
+        assert_eq!(trivial_merge(&[0, 1], &[0, 0, 0]), Some(&0));
         assert_eq!(trivial_merge(&[0, 1], &[0, 0, 1]), Some(&0));
         assert_eq!(trivial_merge(&[0, 1], &[0, 0, 2]), None);
         assert_eq!(trivial_merge(&[0, 1], &[0, 1, 0]), Some(&0));
         assert_eq!(trivial_merge(&[0, 1], &[0, 1, 1]), Some(&1));
         assert_eq!(trivial_merge(&[0, 1], &[0, 1, 2]), Some(&2));
         assert_eq!(trivial_merge(&[0, 1], &[0, 2, 0]), None);
         assert_eq!(trivial_merge(&[0, 1], &[0, 2, 1]), Some(&2));
-        assert_eq!(trivial_merge(&[0, 1], &[0, 2, 2]), None);
+        assert_eq!(trivial_merge(&[0, 1], &[0, 2, 2]), Some(&2));
         assert_eq!(trivial_merge(&[0, 1], &[0, 2, 3]), None);
         assert_eq!(trivial_merge(&[0, 1], &[1, 0, 0]), Some(&0));
         assert_eq!(trivial_merge(&[0, 1], &[1, 0, 1]), Some(&1));
         assert_eq!(trivial_merge(&[0, 1], &[1, 0, 2]), Some(&2));
         assert_eq!(trivial_merge(&[0, 1], &[1, 1, 0]), Some(&1));
```

Comment:
By the way, this is also an ordering issue. It's pretty clear that

```
1    1    0
 ^   ^   ^
  \ / \ /
   1   0
```

should resolve to 1.

Reply:
On third thought, leaving such a conflict ambiguous would do no good. I think the fundamental problem here is as follows, and I don't know how we can resolve it and keep the `A+A-B => A` rule.

I think it would be nice to document this somewhere, but it doesn't have to be in this PR. Also, I'm not sure whether I'll have a better idea of what to say in the future.

--------------------------------

Let's say this is now a portion of a commit tree and we name all the changes:

```
A    B    C
1    1    0
     ^   ^
    /   /
   0   1
   D   E
```

The real problem is that if we do `jj rebase -s C -d B && jj amend -r C && jj rebase -s B -d A`, you get a different result than if you do `jj rebase -s C -d B && jj rebase -s B -d A && jj amend -r C`.

After the first rebase, the value of C is still 0, so you get

```
1
A

0 -> 1 -> 0
D    B    C

1
E
```

#### First option

Now, if we rebase `B` onto `A`, `B` remains a `1` by the `A+A-B => A`. So, `C` remains a 0. In other words, we get

```
1 -> 1 -> 0
A    B    C
```

and after the squash we have

```
1 -> 0
A    B and C squashed
```

#### Second option
If we `jj amend -r C` first, we end up with

```
1
A

0 -> 0
D    B and C squashed

1
E
```

At this point, `B` does not change the value, so rebasing it on top of `A` will result in

```
1 -> 1
A    B and C squashed
```

This is clearly not great, but I have no clue how to do any better since losing the `A+A-B => A` rule is so inconvenient.

Reply:
One potential conclusion from this is that it might end up being useful to keep track which conflict resolutions rely on the `A+A-B => A` rule and which don't. At the moment, that seems more important to me than the discussion about reordering, but even so -- I currently have no use-case for doing it. Keeping track of such information is also not trivial theoretically -- we'd need a model for a conflict that tracks this information in a useful way and can still be rebased on top of any other such conflict in a meaningful way.